File: 프로그래머스/lv3/43163. 단어 변환/단어 변환.py
def check(a, b):
    diff = 0
    for i, j in zip(a, b):
        if i != j:
            diff += 1
    if diff == 1:
        return True
    return False

def dfs(p, target, words, visited):
    global answer
    
    if p == target:
        answer = min(answer, sum(visited))    
        return
    
    for idx, word in enumerate(words):
        if visited[idx]:
            continue
        if not check(p, word):
            continue
        visited[idx] = True
        dfs(word, target, words, visited)
        visited[idx] = False

def solution(begin, target, words):
    visited  = [False]*len(words)
    global answer 
    answer = 1e9
    
    dfs(begin, target, words, visited)
    
    if answer == 1e9:
        return 0
    else:
        return answer

File: 프로그래머스/3/43163. 단어 변환/단어 변환.java
class Solution {
    static boolean[] visited;
    static int answer = (int) 1e9;
    
    public int solution(String begin, String target, String[] words) {
        visited = new boolean[words.length];
        
        dfs(begin, target, words, 0);
        
        if (answer == (int) 1e9) return 0;
        return answer;
    }
    
    public void dfs(String current, String target, String[] words, int cnt) {
        if (current.equals(target)) {
            answer = Math.min(cnt, answer);
            return;
        }
        
        for (int i=0; i<words.length; i++) {
            if(!check(current, words[i])) continue;
            if(visited[i]) continue;
                
            visited[i] = true;
            dfs(words[i], target, words, cnt+1);
            visited[i] = false;
        }        
    }
    
    public boolean check(String a, String b) {
        boolean flag = false;
        for (int i=0; i<a.length(); i++) {
            if (a.charAt(i) != b.charAt(i)) {
                if (flag) {
                    return false;
                }
                flag = true;
            }
        }
        
        if (flag) {
            return true; // 한 자리만 다른 것.
        }
        return false; // 같은 것.
    }
}